lib/lightning_web/components/viewers.ex

```diff
@@ -55,32 +68,117 @@ defmodule LightningWeb.Components.Viewers do
-            class="relative rounded-md p-12 text-center bg-slate-700 font-mono text-gray-200"
-          >
-            <.text_ping_loader>
-              Nothing yet
-            </.text_ping_loader>
+      <div class="relative flex grow">
+        <.log_level_filter
+          :if={@current_user}
+          id={"#{@id}-filter"}
+          levels={@selected_log_levels}
+        />
+
+        <div
+          id={@id}
+          class={["flex grow", @class]}
+          phx-hook="LogViewer"
+          phx-update="ignore"
+          data-run-id={@run_id}
+          data-step-id={@selected_step_id}
+          data-log-levels={Enum.join(@selected_log_levels, ",")}
+          data-loading-el={"#{@id}-nothing-yet"}
+          data-viewer-el={"#{@id}-viewer"}
+        >
+          <div class="relative grow">
+            <div
+              id={"#{@id}-nothing-yet"}
+              class="relative rounded-md p-12 text-center bg-slate-700 font-mono text-gray-200"
+            >
+              <.text_ping_loader>
+                Nothing yet
+              </.text_ping_loader>
+            </div>
+            <div id={"#{@id}-viewer"} class="hidden absolute inset-0 rounded-md">
+            </div>
           </div>
-          <div id={"#{@id}-viewer"} class="hidden absolute inset-0 rounded-md"></div>
         </div>
       </div>
     <% end %>
     """
   end
 
+  defp selected_log_levels(user) do
+    configured_levels = user.preferences["log_levels"] || []
+
+    if configured_levels == [] do
+      all_log_levels()
+    else
+      configured_levels
+    end
+  end
+
+  defp all_log_levels do
+    Lightning.Invocation.LogLine
+    |> Ecto.Enum.values(:level)
+    |> Enum.map(&to_string/1)
+  end
+
+  attr :id, :string, required: true
+  attr :levels, :list
+
+  defp log_level_filter(assigns) do
+    ~H"""
+    <div id={@id} class="absolute top-0 right-4 z-50">
+      <.modal
+        id={"#{@id}-modal"}
+        close_on_click_away={false}
+        close_on_keydown={false}
+      >
+        <:title>
+          <div class="flex justify-between gap-2">
+            <span class="font-bold">
+              Configure log levels
+            </span>
+
+            <button
+              phx-click={hide_modal("#{@id}-modal")}
+              type="button"
+              class="rounded-md bg-white text-gray-400 hover:text-gray-500 focus:outline-none"
+              aria-label={gettext("close")}
+            >
+              <span class="sr-only">Close</span>
+              <.icon name="hero-x-mark" class="h-5 w-5 stroke-current" />
+            </button>
+          </div>
+        </:title>
+        <.form
+          :let={f}
+          id={"#{@id}-form"}
+          for={to_form(%{}, as: :log_filter)}
+          phx-change="save-log-filter"
+        >
+          <fieldset>
+            <div class="space-y-2">
+              <%= for level <- all_log_levels() do %>
+                <.input
+                  type="checkbox"
+                  label={level}
```

Comment:
This is fancy and super flexible, I am just afraid it will become harder to debug because instead of 1 user input we are allowing a permutation of options.

Reply:
It's a comment more about a feature decision not your implementation. I think allowing users to check multiple levels like DEBUG and ERROR makes it very flexible but more complex than rejecting/eliminating all logs that have higher level than the selected. I am not sure how we receive this log levels, if by an integer but would have it as a goal to filter them fastly.

Reply:
Valid point